packages/custom_lint/lib/src/v2/server_to_client_channel.dart

```diff
@@ -90,33 +104,50 @@ void _writePackageConfigForTempProject(
         continue;
       }
 
-      final currentPackage = packageMap[package.name];
+      final currentPackageWithContextRoot = packageMap[package.name];
+      final currentPackage = currentPackageWithContextRoot?.package;
 
       if (currentPackage != null && currentPackage.root != package.root) {
-        throw StateError(
-          '''
-Two ContextRoots depend on ${package.name} but use different version,
-therefore custom_lint does not know which one to pick.
-- ${package.root}
-- ${currentPackage.root}
-''',
+        packagesInMultipleContextRoots.add(
+          _PackageAndContextRootPair(
+            currentPackageWithContextRoot!,
+            _PackageAndContextRoot(package, contextRoot),
+          ),
         );
       }
 
-      packageMap[package.name] = package;
+      packageMap[package.name] = _PackageAndContextRoot(package, contextRoot);
     }
   }
+  if (packagesInMultipleContextRoots.isNotEmpty) {
+    throw StateError(
+      'The following packages are included in multiple ContextRoots:\n'
+      '${[
+        for (final pair in packagesInMultipleContextRoots)
+          '- ${pair.first.package.name}\n'
+              '  -- ContextRoot: ${pair.first.contextRoot.root}\n'
```

Comment:
With this approach, we're likely to have a lot of duplicates in terms of package names.  
Say package A has version conflicts with package B and C, chances are A will be listed twice.

Reply:
true, I can group by package and then show the package only once.